File: ratelimit/tokenbucket.go
package ratelimit

import (
	"sync"
	"time"
)

// tokenBucket implements the token bucket algorithm for rate limiting.
// It maintains a bucket of tokens that refills at a constant rate.
//
// Field alignment optimization is intentionally disabled for this internal struct because:
// 1. This is an internal implementation detail, not exposed in the public API
// 2. Fields are logically grouped (sync primitives, time tracking, token state) for code clarity
// 3. The struct is instantiated once per key, not in hot allocation paths
// 4. Memory overhead from misalignment is negligible (few extra bytes per bucket)
// 5. The mutex and time fields need to be at the top for clear synchronization semantics
//
//nolint:govet // fieldalignment: internal struct, code clarity prioritized over memory optimization
type tokenBucket struct {
	mu sync.Mutex

	// lastRefill is the last time tokens were refilled
	lastRefill time.Time

	// interval is the time period for rate replenishment
	interval time.Duration

	// tokens is the current number of available tokens
	tokens float64

	// burst is the maximum number of tokens (bucket capacity)
	burst float64

	// rate is the number of tokens added per interval
	rate float64
}

// newTokenBucket creates a new token bucket with the given rate, burst, and interval.
func newTokenBucket(rate, burst int, interval time.Duration) *tokenBucket {
	return &tokenBucket{
		tokens:     float64(burst),
		burst:      float64(burst),
		rate:       float64(rate),
		interval:   interval,
		lastRefill: time.Now(),
	}
}

// allow attempts to take 1 token from the bucket.
// Returns true if a token was available, false otherwise.
func (tb *tokenBucket) allow() bool {
	tb.mu.Lock()
	defer tb.mu.Unlock()

	tb.refill()

	if tb.tokens >= 1.0 {
		tb.tokens--
		return true
	}

	return false
}

// take attempts to take n tokens from the bucket.
// Returns true if n tokens were available, false otherwise.
// If insufficient tokens are available, no tokens are taken.
func (tb *tokenBucket) take(n int) bool {
	tb.mu.Lock()
	defer tb.mu.Unlock()

	tb.refill()

	tokensNeeded := float64(n)
	if tb.tokens >= tokensNeeded {
		tb.tokens -= tokensNeeded
		return true
	}

	return false
}

// waitTime returns the duration to wait for at least 1 token to become available.
// Returns 0 if a token is immediately available.
//
// IMPORTANT: This is a best-effort calculation. The returned wait time is based on
// the current token count and refill rate, but is NOT a guarantee. Between calculating
// the wait time and the caller using it, other goroutines may consume tokens, making
// the wait time stale. The caller should be prepared to wait again if needed.
//
// Includes safeguards against edge cases like zero rate or very small intervals.
func (tb *tokenBucket) waitTime() time.Duration {
	tb.mu.Lock()
	defer tb.mu.Unlock()

	tb.refill()

	// If tokens available, no wait needed
	if tb.tokens >= 1.0 {
		return 0
	}

	// Calculate how many tokens we need
	tokensNeeded := 1.0 - tb.tokens

	// Safety check: ensure tokensNeeded is positive
	if tokensNeeded <= 0 {
		return 0
	}

	// Safety check: if rate is 0 or negative, tokens will never be added
	// Return maximum wait time to prevent infinite blocking
	if tb.rate <= 0 {
		return time.Hour * 24
	}

	// Calculate time to generate those tokens
	// tokens per nanosecond = rate / interval
	intervalNs := tb.interval.Nanoseconds()

	// Safety check: prevent division by zero or very small intervals
	if intervalNs <= 0 {
		return time.Hour * 24 // Maximum wait for edge case
	}

	tokensPerNs := tb.rate / float64(intervalNs)

	// Safety check: if tokens per nanosecond is effectively zero (very small rate)
	// use direct calculation to avoid division by zero
	if tokensPerNs <= 0 {
		return time.Hour * 24
	}

	nsToWait := tokensNeeded / tokensPerNs

	// Safety check: ensure result is within reasonable bounds
	// Prevent negative durations and cap maximum wait time
	if nsToWait < 0 {
		return 0
	}

	wait := time.Duration(nsToWait)

	// Cap maximum wait time to 24 hours to prevent overflow and unreasonable waits
	maxWait := time.Hour * 24
	if wait > maxWait {
		return maxWait
	}

	return wait
}

// refill adds tokens to the bucket based on time elapsed since last refill.
// Must be called with lock held.
func (tb *tokenBucket) refill() {
	now := time.Now()
	elapsed := now.Sub(tb.lastRefill)

	if elapsed <= 0 {
		return
	}

	// Cap elapsed time to prevent overflow from clock skew or system sleep
	// Maximum 1 hour of elapsed time is reasonable for token refill
	if elapsed > time.Hour {
		elapsed = time.Hour
	}

	// Calculate tokens to add: (elapsed / interval) * rate
	tokensToAdd := (float64(elapsed.Nanoseconds()) / float64(tb.interval.Nanoseconds())) * tb.rate

	tb.tokens += tokensToAdd
	if tb.tokens > tb.burst {
		tb.tokens = tb.burst
	}

	tb.lastRefill = now
}

File: backends/redis/lua/token_bucket.lua
--[[
Token Bucket Rate Limiting Script for Redis

This script implements atomic token bucket rate limiting with automatic refill.
It handles token refill based on elapsed time and attempts to consume tokens in a single atomic operation.

KEYS[1]: bucket key (e.g., "fortify:ratelimit:user:123")

ARGV[1]: rate (number of tokens added per interval)
ARGV[2]: burst (maximum tokens in bucket)
ARGV[3]: interval (nanoseconds between token additions)
ARGV[4]: tokens_to_take (number of tokens to consume, typically 1)
ARGV[5]: current_time_ns (current timestamp in nanoseconds)
ARGV[6]: ttl_seconds (bucket TTL in seconds for auto-cleanup)

Returns:
  1 = allowed (tokens available)
  0 = denied (insufficient tokens)

Data Structure:
  The bucket is stored as a Redis hash with two fields:
  - tokens: current number of available tokens (float)
  - last_refill: last refill timestamp in nanoseconds (integer)
--]]

local bucket_key = KEYS[1]

local rate = tonumber(ARGV[1])
local burst = tonumber(ARGV[2])
local interval_ns = tonumber(ARGV[3])
local tokens_to_take = tonumber(ARGV[4])
local current_time_ns = tonumber(ARGV[5])
local ttl_seconds = tonumber(ARGV[6])

-- Get current bucket state
local bucket = redis.call('HMGET', bucket_key, 'tokens', 'last_refill')
local current_tokens = tonumber(bucket[1])
local last_refill_ns = tonumber(bucket[2])

-- Initialize new bucket if it doesn't exist
if not current_tokens or not last_refill_ns then
    current_tokens = burst
    last_refill_ns = current_time_ns
end

-- Calculate elapsed time since last refill
local elapsed_ns = current_time_ns - last_refill_ns

-- Safety check: prevent negative elapsed time (clock skew)
if elapsed_ns < 0 then
    elapsed_ns = 0
end

-- Cap elapsed time to prevent overflow (max 1 hour)
local max_elapsed_ns = 3600 * 1000000000  -- 1 hour in nanoseconds
if elapsed_ns > max_elapsed_ns then
    elapsed_ns = max_elapsed_ns
end

-- Calculate tokens to add based on elapsed time
-- tokens_to_add = (elapsed / interval) * rate
local tokens_to_add = 0
if interval_ns > 0 and rate > 0 then
    tokens_to_add = (elapsed_ns / interval_ns) * rate
end

-- Refill tokens
current_tokens = current_tokens + tokens_to_add

-- Cap at burst limit
if current_tokens > burst then
    current_tokens = burst
end

-- Attempt to take tokens
local allowed = 0
if current_tokens >= tokens_to_take then
    current_tokens = current_tokens - tokens_to_take
    allowed = 1
end

-- Update bucket state
redis.call('HSET', bucket_key,
    'tokens', tostring(current_tokens),
    'last_refill', tostring(current_time_ns))

-- Set TTL to auto-expire unused buckets
redis.call('EXPIRE', bucket_key, ttl_seconds)

return allowed
